collaborative-pre-merge-testing-for-multi-PR-features/README.md

```diff
@@ -0,0 +1,580 @@
+
+```html
+{{if.surcharge}}
+<div class="surcharge-info" style="padding: 10px; background-color: #fffbe6; border: 1px solid #ffe58f; margin-top: 15px; border-radius: 5px;">
+    <strong>Dynamic Surcharge Applied:</strong> ${{printf "%.2f".surcharge}}
+</div>
+{{end}}
+
+<div id="ride-info" class="p-2"></div>
+```
+
+### 2.3 Build New Docker Images
+
+**When you create the sandboxes, they need to use a new image - the one that has the code changes for the new feature.**
+
+After making the code changes, build new container images:
+
+```bash
+# Generate Go code from proto files
+cd hotrod
+make generate-proto
+
+# Build the React frontend
+make build-frontend-app
+
+# Build the Docker image with all changes
+make build-docker
+
+# Tag and push the image
+docker tag hotrod:latest <your-dockerhub-username>/hotrod:epic-42
+docker push <your-dockerhub-username>/hotrod:epic-42
+```
+
+**Note:** Replace `<your-dockerhub-username>` with your actual Docker Hub username or registry path.
+
+---
+
+## 3. Manual Workflow: Unified Preview with Signadot
+
+### 3.1. Create Sandboxes for Each Service
+
+Create `route-surcharge-sandbox.yaml`:
+
+```yaml
+name: route-surcharge-feature
+spec:
+  cluster: <your-cluster-name>
+  labels:
+    epic: EPIC-42
+  forks:
+    - forkOf:
+        kind: Deployment
+        namespace: hotrod
+        name: route
+      customizations:
+        images:
+          - image: <your-dockerhub-username>/hotrod:epic-42
+```
+
+Create `frontend-surcharge-sandbox.yaml`:
+
+```yaml
+name: frontend-surcharge-feature
+spec:
+  cluster: <your-cluster-name>
+  labels:
+    epic: EPIC-42
+  forks:
+    - forkOf:
+        kind: Deployment
+        namespace: hotrod
+        name: frontend
+      customizations:
+        images:
+          - image: <your-dockerhub-username>/hotrod:epic-42
+        env:
+          - name: FRONTEND_ROUTE_ADDR
+            value: "route-surcharge-feature-dep-route-<unique-id>.hotrod.svc:8083"
+```
+
+Apply the sandboxes:
+
+```bash
+signadot sandbox apply -f route-surcharge-sandbox.yaml --set cluster=<your-cluster-name>
+signadot sandbox apply -f frontend-surcharge-sandbox.yaml --set cluster=<your-cluster-name>
+```
+
+**Important:** In your cluster, there will be 2 versions of the frontend and route deployments (baseline & sandbox).
+
+**Troubleshooting:**
+
+- If you see `Insufficient cpu` errors, increase Minikube resources (see above). The Signadot traffic manager sidecars require significant CPU.
+- If you see `ImagePullBackOff` for custom images, ensure the image exists in your registry or use the baseline image.
+- If pods are stuck in `Pending`, describe them:
+
+  ```bash
+  kubectl -n hotrod describe pod <pod-name>
+  ```
+
+  Look for resource or image pull errors.
+
+### 3.2. Create a RouteGroup
+
+Create `epic-routegroup.yaml`:
+
+```yaml
+name: epic-42-preview
+spec:
+  cluster: <your-cluster-name>
+  match:
+    any:
+      - label:
+          key: epic
+          value: EPIC-42
+  endpoints:
```

Comment:
These endpoints are not needed. You can just hit the regular http://frontend.hotrod.svc:8080/.